### PR TITLE
[SPARK-14030][MLLIB] Add parameter check to MLLIB

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?

add parameter verification to MLLIB, like
numCorrections > 0
tolerance >= 0
iters > 0
regParam >= 0
## How was this patch tested?

manual tests

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala [hunk 1]
```diff
@@ -52,7 +52,8 @@ class LBFGS(private var gradient: Gradient, private var updater: Updater)
    * Restriction: numCorrections > 0
    */
   def setNumCorrections(corrections: Int): this.type = {
-    assert(corrections > 0)
+    require(corrections > 0, s"Number of corrections must be greater than 0," +
+      s" but got ${corrections}")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala]: Would the following style fit? If yes, this is slightly easier to read.

~~~scala
require(corrections > 0,
  s"Number of corrections must be greater than 0 but got ${corrections}")
~~~

File: mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala [hunk 2]
```diff
@@ -88,6 +91,8 @@ class LBFGS(private var gradient: Gradient, private var updater: Updater)
    * Set the maximal number of iterations for L-BFGS. Default 100.
    */
   def setNumIterations(iters: Int): this.type = {
+    require(iters > 0, s"Maximum of iterations must be greater than 0," +
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala]: In some cases, we might need `iters = 0` to test baseline models.

**Comment by srowen** [comment]
@zhengruifeng let's not open a separate issue for every file where you want to add args checking. Combine these into one JIRA and PR and close the others.

File: mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala (hunk 1 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala]: ok, I will reformat all args-checking like this.

File: mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala (hunk 2 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala]: ok, I will fix it.

**Comment by zhengruifeng** [comment]
@srowen  What about having only one JIRA with several PRs? For there may be many kinds of algorithm to add args-checking: such as classification, clustering...

**Comment by srowen** [comment]
Why? these are tiny and logically related changes. Just survey all similar classes/methods at once and add similar checks where appropriate. It makes less sense to do it half-way.

**Comment by zhengruifeng** [comment]
@srowen Just because there are many files to deal with, and some parameter may need some time or discuss to determine the range.
I will first add the parameters which are obvious to check into this PR.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/NaiveBayes.scala [hunk 3]
```diff
@@ -326,6 +326,8 @@ class NaiveBayes private (
   /** Set the smoothing parameter. Default: 1.0. */
   @Since("0.9.0")
   def setLambda(lambda: Double): NaiveBayes = {
+    require(lambda > 0,
+      s"Smoothing parameter must be greater than 0 but got ${lambda}")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/NaiveBayes.scala]: Yeah this is looking good, to get all of the same obvious parameter checks done in one pass. You can say "positive" for "greater than 0" and "nonnegative" for "greater than or equal to 0" but this is fine.

**Comment by srowen** [comment]
LGTM

File: mllib/src/main/scala/org/apache/spark/mllib/classification/NaiveBayes.scala (hunk 3 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/NaiveBayes.scala]: ok， I will replace them for short

File: mllib/src/main/scala/org/apache/spark/mllib/classification/NaiveBayes.scala [hunk 4]
```diff
@@ -326,6 +326,8 @@ class NaiveBayes private (
   /** Set the smoothing parameter. Default: 1.0. */
   @Since("0.9.0")
   def setLambda(lambda: Double): NaiveBayes = {
+    require(lambda > 0,
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/NaiveBayes.scala]: Oops, one more thing -- lambda = 0 should be allowed as it corresponds to no additive smoothing.

Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/NaiveBayes.scala]: ok, I will fix it

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**